### ISSUE TITLE
Poor performance with Atom v1.21.2

### ISSUE THREAD

**Issue by DavidLGoldberg** [DavidLGoldberg/jumpy#108]
The release of [Atom v1.21.2](https://github.com/atom/atom/releases/tag/v1.21.2) has rendered Jumpy *almost* useless.  The good news is there should be a fix coming in the next patch/minor release!  The dev build of Atom no longer seems to have this issue.  I think the developers said this is coming soon (this week).  Thanks @nathansobo and @leroix for the quick fixes :)

Unfortunately this still affects the 4.0.0 release of Jumpy (still slow).

The initial PR:
https://github.com/atom/atom/pull/15894

Issue raised:
https://github.com/atom/atom/issues/16077

Fix PR:
https://github.com/atom/atom/pull/16083

As noted this should get launched soon.  I don't think there is much I can / should do until then.  Thanks!

**Comment by DavidLGoldberg** [comment]
Toying with the idea of trying to get a baseline performance / timing test in Travis CI, but I don't think a good solution exists yet...

There might be some overhauls to the decorations I can do with layers that would improve performance / prevent this stuff in the future?

**Comment by DavidLGoldberg** [comment]
Hmm, so after the 1.22.0 release....the initial labeling is fast, but it takes forever for the first character to be entered (which behind the scenes triggers the adding of a class to most of the decorations to hide the non matching ones)

The 2nd character's response is faster (although there are fewer labels/decorations still on the screen 26 usually).....the jump and beacon animation then happens fast etc.

Still works fast in the dev build :\
@nathansobo, @leroix

Is there a change to the decoration architecture that requires me to rewrite this stuff?  Do I *need* layers etc?  Although, anytime I do `atom -d` or use the dev build it all works fast as it used to.

**Comment by nathansobo** [comment]
@leroix We may need to more aggressively backport some of your recent changes to resize. We need to keep investigating this.

**Comment by leroix** [comment]
@nathansobo @DavidLGoldberg This seems to be because after the first character is typed and irrelevant markers are destroyed, a resize is getting triggered on all the `OverlayComponent`s that should be getting destroyed. In 1.23, it's just a whole lot more efficient to resize an `OverlayComponent` since it doesn't invoke update on the entire editor.

For some reason, this doesn't happen when you just exit jumpy. i.e. hit escape or enter after hitting shift-enter. I'm looking through jumpy to see what might be different in the two scenarios to try to figure out why one results in tons of resizes and the other does not.

Bottom line: destroying markers shouldn't be triggering `OverlayComponent` resizes, so there does seem to be a bug in Atom.

Here's Atom 1.24.0-dev-6227ece when I hit the first character. It results in a ton of calls to the `ResizeObserver` callback. In 1.22, each one of these calls is way more expensive.
![screen shot 2017-11-07 at 9 31 03 pm](https://user-images.githubusercontent.com/520209/32531755-fd0afea8-c402-11e7-970e-782fee494602.png)

**Comment by leroix** [comment]
@DavidLGoldberg @nathansobo Ah, looking a little closer at jumpy's source. It seems that it sets `display: none` on the irrelevant labels. Could jumpy destroy them instead?

It seems like destroyed markers aren't resulting in resizes after all.

**Comment by leroix** [comment]
Oh I see. `backspace` brings them back. Hmm...

**Comment by leroix** [comment]
For context, the change that we made in Atom that caused this was https://github.com/atom/atom/pull/15894.

We started observing `OverlayComponent`'s child for resizes instead of `OverlayComponent` itself because folks can change the child in strange ways that aren't necessarily reflected in the size of `OverlayComponent`.

**Comment by DavidLGoldberg** [comment]
Thanks for more hard work!  Yes, there were a few reasons I went with the class add at the time.  Seemed quick / efficient than dom manip....thoughts?  The backspace to and from is SOO much faster than repainting...

**Comment by DavidLGoldberg** [comment]
I learned that pretty early on, before decorations were a thing (or at least before I was using them) when I was doing my own dom elements....... Honestly so one thing that I've always thought about is..........is one dom element at a higher z level ....that uses absolute positioning would *probably* (untested) be faster for the rendering than attaching hundreds of dom elements......btw.....all the lag that you can see from larger text files or multiple large text files (across panes) ... all of the lag seems to come from rendering as opposed to other processing / regex etc...(From what I can tell with my experiments)

**Comment by DavidLGoldberg** [comment]
Which is what lead me to wonder if that's what any of the layering stuff does...I haven't looked *too* much into the atom rendering tech lately, so kind of ignorant.  I know at one point I had to abandon my dom manip and jump on the decoration bandwagon.  I forget the exact details but the 2 competing processes were colliding, so rather than make 2 calls to *something* I used the decorations that had that call for free....I forget the specifics.  I was deep down in the inspector stack.

BUT if I just tack on ONE big dom element with a high z-index opacity set to clear for non child elements... maybe?

**Comment by DavidLGoldberg** [comment]
One thing I've never been able to seem to defeat, is the decorations seem to leave the green "node" line in inspector always rising!

I can't GC to lower it back down.  I had some more noticeable issues that I was able to improve the GC on for example..

Recently though, I had it isolated to a call to the decorations, creating the green hill.

I've followed the docs on how to destroy by destroying the *marker* (right)?

If there is a dangerous leak, it probably has nothing to do with this current issue though :\  But wanted to throw it out there in case it helps indicate I'm doing something else wrong (ish).

**Comment by leroix** [comment]
Decorators are actually pretty efficient, and they're only getting better in the latest Atom. Here's how long it takes to render and paint all of jumpy's labels in the latest Atom. It's well under 100ms which is great in this case.
![screen shot 2017-11-07 at 10 30 41 pm](https://user-images.githubusercontent.com/520209/32533021-6977f110-c40b-11e7-9fb2-e0c2510b4381.png)

So, setting a new class (`.irrelevant`) on labels still results in a layout flow:
![screen shot 2017-11-07 at 10 32 39 pm](https://user-images.githubusercontent.com/520209/32533067-ab68f2c2-c40b-11e7-85c7-2ce4eef757d3.png)

Here's how long it takes to remove all the labels:
![screen shot 2017-11-07 at 10 34 03 pm](https://user-images.githubusercontent.com/520209/32533090-c7f5f4e4-c40b-11e7-8420-2afb27ae5a91.png)

Yea, I see what you're saying about nodes always rising. I'm not sure why that would be though. I think you're destroying markers correctly.

**Comment by leroix** [comment]
Granted, in that second screenshot, you have a lot of time taken up by those resize events.

I think the solution, if you want to keep setting the class rather than removing the labels, may be to ensure that the label's size stays constant. Try `visibility: hidden` perhaps?

**Comment by DavidLGoldberg** [comment]
So in the current release, adding the irrelevant, gotcha.  Kind of gives me a crazy idea of putting the irrelevant on and off a parent class for each, as opposed to manipulating the sub elements and letting specificity do it's thing (hah!)  Obviously removing the decoration would be ideal ;)

**Comment by DavidLGoldberg** [comment]
Well just a quick reminder, in the newest build everything is super fast.  Unless I'm using some wrong build...?

**Comment by DavidLGoldberg** [comment]
Like, someone seems to have fixed it (double checking atom-beta now)

**Comment by leroix** [comment]
yea, definitely try `visibility: hidden` before you spend a bunch of time changing it to add and remove dom elements.

It's _faster_, but it's still doing a lot of work in those resize observer callbacks that it doesn't need to be. e.g. second screenshot.

**Comment by DavidLGoldberg** [comment]
Jumpy was never *completely* instant (wish it was).  But recently especially with recent atom changes it was definitely fast enough to not care.

**Comment by DavidLGoldberg** [comment]
like the use case I care about is....see the labels, spurt out 2 keys and it jumps *real* fast.... so the irrelevant's are quick....Just did that on `atom-beta`

**Comment by DavidLGoldberg** [comment]
so like:
```
    &.irrelevant { // This is a marker to indicate not applicable (when 1 key is entered/known).
        display: none;
        visibility: hidden;
    }
```

**Comment by leroix** [comment]
I think for jumpy, as it's a response to a user action, as long as you stay well under that 100ms threshold you're gravy.

Yea, in beta, resizes of the `OverlayComponent` only result in an update of the `OverlayComponent` itself. In 1.22 and prior, it updates the entire text editor component.

**Comment by leroix** [comment]
You should be able to replace `display: none` with `visibility: hidden` right?

**Comment by DavidLGoldberg** [comment]
decent heavy text files
on my 4 year old work mac (I know time for a new one, getting one soon) It definitely goes above that.

At some point I had to remove the large_text.md spec test.....because it takes too long (not sure what happened there, but the real use case isn't as bad...but still slow

**Comment by DavidLGoldberg** [comment]
I seem to remember there being some need for display: none i could be wrong though...

**Comment by DavidLGoldberg** [comment]
ok, let me try non beta and that change give me a second

**Comment by DavidLGoldberg** [comment]
see so..... the problem is.....all of my -d env's are fast.

**Comment by DavidLGoldberg** [comment]
Any insight into that problem...?

**Comment by leroix** [comment]
`visibility: hidden` just sets the opacity of the element to 0 (and keeps its size).

Checking on the dev mode of 1.22.

**Comment by DavidLGoldberg** [comment]
yes, please :)

**Comment by leroix** [comment]
Oh, dev mode runs code from your local Atom repo. Even if you started up 1.22.

**Comment by DavidLGoldberg** [comment]
Of course this could totally be me breaking my dev env with that local build....although that went relatively smoothly for first time....

By the way you're not a github employee right?  Thanks for all of the crazy work!

**Comment by DavidLGoldberg** [comment]
Sure, but it has no changes....so I've been using dev mode for like 3 years now ish and never seen this heh.

**Comment by DavidLGoldberg** [comment]
So like, 
```
➜  jumpy git:(master) ✗ git stash save
Saved working directory and index state WIP on master: e6af0d8 Prepare 4.1.0 release
HEAD is now at e6af0d8 Prepare 4.1.0 release
```

**Comment by leroix** [comment]
So, since it's running from your local Atom repo, it's benefiting from those recent changes to make overlay resizes faster.

I'm just a guy that wants to make Atom great :P. Jumpy's a cool package, and it looks like a ton of people are using it.

**Comment by DavidLGoldberg** [comment]
So,
```
➜  jumpy git:(master) gst
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working tree clean```

**Comment by DavidLGoldberg** [comment]
and yet the behavior is different :\

**Comment by DavidLGoldberg** [comment]
So, that's kind of where I'm confused, how does it get those new recent changes?

**Comment by leroix** [comment]
Yea, I think maybe I'm not being clear. The Atom in dev mode (even 1.22) runs Atom code from your local Atom repo. By default, it looks at ~/github/atom.

**Comment by DavidLGoldberg** [comment]
In the past, I thought I had to go to atom-beta for that kidn of thing....and it wasn't until I installed the atom dev build (I'll call it baby blue ;) ) that this behavior started happening....I think it messed with my file system....unless....your's also has that behavior....like atom vs atom -d ....the ONLY changes should be the dev packages under test (to my understanding)  (?????)

**Comment by DavidLGoldberg** [comment]
~/github/atom !!!!! ah! let me think about that a bit more

**Comment by leroix** [comment]
Yea, `atom -d` always runs from your local atom repo. Since you've got a recent commit on master checked out, you probably get the change we made in https://github.com/atom/atom/pull/15954 which makes overlay resizes a lot faster.

**Comment by DavidLGoldberg** [comment]
oh no way is ~/github/atom "a thing".....i *happen* to use ~/github (did you just deduce that's what i'm using)?

**Comment by DavidLGoldberg** [comment]
yeah so that atom -d always going there..........kind of....dare I say it.....*not* a fan?

**Comment by leroix** [comment]
lol, that happened to me too. Atom just happens to look there by default. There's an environment variable that's something like ATOM_DEV_PATH that you can set if it's somewhere else.

**Comment by DavidLGoldberg** [comment]
at the current moment I have no way to test atom -d on I'll call it "green" atom stable.

**Comment by DavidLGoldberg** [comment]
You're the man, I had some of this figured out but you connected the dots.

**Comment by leroix** [comment]
You could just check out the tag and reload Atom.

**Comment by leroix** [comment]
Awesome! Happy to help :)

**Comment by DavidLGoldberg** [comment]
so once I get my -d fixed.  I can test the visibility fix...which should get me to a usable pre launch fix.

**Comment by DavidLGoldberg** [comment]
In theory right?

**Comment by leroix** [comment]
You mean if you `git checkout v1.22.0` and run `atom -d`? Yea, I believe so.

**Comment by DavidLGoldberg** [comment]
Oh, I have to rebuild, don't I (oops)

**Comment by DavidLGoldberg** [comment]
I'm doing
`script/build --install`

**Comment by DavidLGoldberg** [comment]
I had an atom, where like I could navigate around the tree view, but not click a file to open or hit enter (twilight zone)

**Comment by DavidLGoldberg** [comment]
honestly, might just be easier to test in prod ;)

jk almost there I think.

**Comment by DavidLGoldberg** [comment]
maybe the downloadble again and mv ing atom out or.... the env var is a quicker approach heh.

**Comment by leroix** [comment]
Weird o_0 (about the tree view). It should work to download 1.22 from atom.io, `get fetch --tags`, `git checkout v1.22.0`, `atom -d`.

Rebuilding works too though.

**Comment by DavidLGoldberg** [comment]
I abandoned the build, was taking a while (I might have left the wrong configs of atom's opened while it started too, not sure if that's a thing or not)... so I mv ~/github/atom ~/github/atombak

Re ran the installer (heh) and now i'm on green stable BOTH broke atom / atom -d  (joyous day!) now let me test this change!

**Comment by DavidLGoldberg** [comment]
You are a GOD.  That was REALLY clever, thanks.  I was still kind of thinking it was related to the first issue, but of course obviously very different!  Thanks man really appreciate it.  I'll push a fix right now, remove the notification too.

**Comment by leroix** [comment]
🍾  🎉  Yay! So glad that fixes it. Feel free to reach out any time. :)

**Comment by DavidLGoldberg** [comment]
btw, fyi: some guy copied it a while back..it's not full featured...now that I ported it to typescript...going to get it in there as well :)
https://marketplace.visualstudio.com/items?itemName=wmaurer.vscode-jumpy


Although I want to try a round of elm for some core shared components.....did you see those "labelers" in the code? that's where most of the atom interface is ...so I could build those for vs code in theory too...

**Comment by DavidLGoldberg** [comment]
Yeah, likewise if there's ever anything I can help with :\

Hah!

**Comment by DavidLGoldberg** [comment]
Oh one last thing, do you have a discuss.atom.io name?
